sunspec.go

```diff
@@ -28,10 +28,125 @@ type ScaleFactor int16
 
 type Count uint16
 
+// An array is a collection of devices.
+//
+// This type is used when the underlying representation
+// is an XML document.
+type Array interface {
+	Device(id string) Device // Answer the specified device instance or nil
+	Do(func(d Device))       // Iterate over all the devices.
+}
+
+// A Device is a collection of Models.
+type Device interface {
+	Model(i ModelId) Model // Answer the specified model instance or nil
+	Do(func(m Model))      // Iterate over all the models supported by the device
+}
+
+// A Model is a collection of Blocks.
 type Model interface {
-	GetId() ModelId
+	GetId() ModelId // Answer the model identifier of the receiver
+	Blocks() int    // Answer the number of blocks
+
+	Block(i int) Block // Answer the nth Block.
+	Do(func(b Block))  // Iterate across all the blocks, in order.
 }
 
-type Device struct {
-	Models []Model
+// A Block is a collection of Points
+//
+// A Block maintains a buffer containing a copy of parts of the device address space,
+//
+// The Read operation populates the buffer from the physical device (using modbus or some other protocol)
+// while a Write operation writes the buffer into the the physical device.
+//
+type Block interface {
+	Point(id string) Point // Answer the specified point, or an error point otherwise
+	Do(func(p Point))      // Iterate over all the points in the block
+
+	// Read the specified pointIds or all if none is specified. A successful read implies
+	// that subsequent access to the specified points will be error free.
+	Read(pointIds ...string) error
+
+	// Write the specified pointIds (or all, if none specified) into the physical device.
+	Write(pointIds ...string) error
+}
+
+// A Point is collection of one or more 16 bit registers that is mapped
+// onto a strongly typed golang type of the correct shape.
+//
+// The value associated with a Point must be accessed by the properly typed
+// accessor or a panic will result. An attempt to obtain the value of a point
+// which has a current error or which has not yet been read from the physical device
+// will also result in a panic.
+//
+// Points also know how to apply scaling factors if an appropriate accessor
+// method is used, so that if you use the Float64() method with a 16 or 32 bit
+// register for which a scaling factor is defined, then the register will be
+// multiplied by the scaling factor to produce the expected 64 bit float result.
+// Note that both the point and the related scaling factor point (if any) must
+// be error free at the point the scaling factor is applied or a panic will result.
```

Comment:
The Point type is designed to be type safe and not fiddly to use (e.g. no error checking on each any every access to a point).  The way that it is achieves this is to keep track of both a value and an error. If the error is not nil, then the point is in an error state and any attempt to use its value will result in a panic. A point starts out in an uninitialised error state and moves to an error free state when a value is associated with the point, usually after a Read.

Provided you have done enough to ensure that the point is present (by issuing a successful Read call) and you choose the correctly typed accessor method, it will not panic. This is the "error free" case. 

However, if you fail to do these things then it will panic.  In particular if you try to get the scaled float64 value of a point that has a scale factor, then it will panic if either the point is in an error state or it's related scale factor point is in an error state, perhaps because it hasn't been read yet.